src/ucs/stats/serialization.c

```diff
@@ -303,22 +303,58 @@ ucs_stats_serialize_binary(FILE *stream, ucs_stats_node_t *root,
 }
 
 static ucs_status_t
-ucs_stats_serialize_text_recurs(FILE *stream, ucs_stats_node_t *node,
-                                ucs_stats_children_sel_t sel, unsigned indent)
+ucs_stats_serialize_text_recurs_filtered(FILE *stream, ucs_stats_filter_node_t *filter_node,
+                                         unsigned indent)
 {
-    ucs_stats_node_t *child;
+    ucs_stats_filter_node_t *filter_child;
+    ucs_stats_node_t *node;
     unsigned i;
-
-    fprintf(stream, "%*s"UCS_STATS_NODE_FMT":\n", indent * 2, "",
-            UCS_STATS_NODE_ARG(node));
-
-    for (i = 0; i < node->cls->num_counters; ++i) {
-        fprintf(stream, "%*s%s: %"PRIu64"\n", (indent + 1) * 2, "",
-                node->cls->counter_names[i], node->counters[i]);
+    int is_sum = ucs_global_opts.stats_format == UCS_STATS_SUMMARY;
+    char *nl = is_sum ? "" : "\n";
+    char *space =  is_sum ? "" : " ";
+    char *left_b = is_sum ? "{" : "";
+    char *rigth_b = is_sum ? "} " : "";
+
+    if (!filter_node->ref_count) {
+        return UCS_OK;
     }
-
-    ucs_list_for_each(child, &node->children[sel], list) {
-        ucs_stats_serialize_text_recurs(stream, child, sel, indent + 1);
+    if (filter_node->type_list_len > 1) {
+        fprintf(stream, "%*s%s*:%s", UCS_STATS_INDENT(is_sum, indent),
+                filter_node->cls->name, nl);
+    } else {
+        fprintf(stream, "%*s"UCS_STATS_NODE_FMT":%s",
+                UCS_STATS_INDENT(is_sum, indent),
+                UCS_STATS_NODE_ARG(filter_node), nl);
+    }
+    // Root shouldn't be with brackets.
+    if (filter_node->parent) {
+        fprintf(stream, "%s", left_b);
+    }
+    for (i = 0; i < filter_node->cls->num_counters; ++i) {
+        uint64_t counters_acc = 0;
+        if (filter_node->counters_bitmask & (1<<i)) {
+            ucs_list_for_each(node, &filter_node->type_list, type_list) {
+                counters_acc += node->counters[i];
+            }
+            fprintf(stream, "%*s%s%s:%s%"PRIu64"%s",
+                    UCS_STATS_INDENT(is_sum, indent + 1),
+                    filter_node->cls->counter_names[i],
+                    filter_node->type_list_len > 1 ? "*" : "",
+                    space, counters_acc, nl);
+            /* Don't print space on last counter */
+            if (UCS_STATS_IS_LAST_COUNTER(filter_node->counters_bitmask, i) &&
+                is_sum) {
+                fprintf(stream, " ");
+            }
+        }
+    }
+    ucs_list_for_each(filter_child, &filter_node->children, list) {
+        ucs_stats_serialize_text_recurs_filtered(stream, filter_child,
+                                                 indent + 1);
+    }
+    if (filter_node->parent) {
+        // Root shouldn't be with parent brackets.
+        fprintf(stream, "%s", rigth_b);
```

Comment:
i suggest to use fputs() when writing unformatted string, should be slightly faster